src/Codecs.chpl

```diff
@@ -0,0 +1,110 @@
+module Codecs {
+  use CTypes;
+  use idna;
+  use iconv;
+
+  use AryUtil;
+  
+  proc encodeStr(obj: c_ptr(uint(8)), inBufSize: int, outBufSize: int, toEncoding: string = "UTF-8", fromEncoding: string = "UTF-8"): [] uint(8) throws {
+    if toEncoding == "IDNA" {
+      if outBufSize == 1 {
+        var ret: [0..0] uint(8) = 0x00;
+        return ret;
+      }
+      var cRes: c_string;
+      var rc = idn2_to_ascii_lz(obj:c_string, cRes, 0);
+      if (rc != IDNA_SUCCESS) {
+        throw new Error("Encode failed");
+      }
+      var tmp = cRes: c_ptr(uint(8));
+      var ret: [0..#outBufSize] uint(8);
+      for i in ret.domain do ret[i] = (tmp+i).deref();
+      idn2_free(cRes: c_void_ptr);
+      return ret;
+    } else if fromEncoding == "IDNA" {
+      if outBufSize == 1 {
+        var ret: [0..0] uint(8) = 0x00;
+        return ret;
+      }
+      var cRes: c_string;
+      var rc = idn2_to_unicode_8z8z(obj:c_string, cRes, 0);
```

Comment:
It looks like if `fromEncoding` is `"IDNA"`, we convert to UTF-8 regardless of what `toEncoding` is. Is that the right thing to do if the `toEncoding` is something other than `"UTF-8"`? Similarly the situation where `toEncoding` is `"IDNA"`, we don't check what `fromEncoding` is set to.

It's fine if we don't support idna encode/decode with anything other than UTF-8, but I feel like we should check what encoding is passed and error out when not UTF-8.

Reply:
Ya, this is something that I had thought about as well. I think that in the short-term, maybe we should error out, like you said, but we also could make it a two-pass implementation where we convert from idna -> utf-8 -> requested encoding, but that could already be achieved through a slightly less convenient client-side interaction.

Does that sound good to just error for this initial effort?

Reply:
oh yeah the two pass is clever! I'm happy with the error for the initial impl